desc/objectives/_free_boundary.py

```diff
@@ -0,0 +1,882 @@
+
+        self._constants = {
+            "eval_transforms": eval_transforms,
+            "eval_profiles": eval_profiles,
+            "src_transforms": src_transforms,
+            "src_profiles": src_profiles,
+            "interpolator": interpolator,
+            "ext_field": self._ext_field,
+            "quad_weights": np.sqrt(np.tile(eval_transforms["grid"].weights, neq)),
+        }
+
+        if self._sheet_current:
+            sheet_current = self.things[1]
+            assert (
+                (sheet_current.M == eq.surface.M)
+                and (sheet_current.N == eq.surface.N)
+                and (sheet_current.NFP == eq.surface.NFP)
+                and (sheet_current.sym == eq.surface.sym)
+            ), "sheet current must have same resolution as eq.surface"
+            self._sheet_data_keys = ["K"]
+            sheet_eval_transforms = get_transforms(
+                self._sheet_data_keys, obj=sheet_current, grid=eval_grid
+            )
+            sheet_src_transforms = get_transforms(
+                self._sheet_data_keys, obj=sheet_current, grid=src_grid
+            )
+            self._constants["sheet_eval_transforms"] = sheet_eval_transforms
+            self._constants["sheet_src_transforms"] = sheet_src_transforms
+
+        timer.stop("Precomputing transforms")
+        if verbose > 1:
+            timer.disp("Precomputing transforms")
+
+        self._dim_f = neq * eval_grid.num_nodes
+
+        if self._normalize:
+            scales = compute_scaling_factors(eq)
+            self._normalization = scales["B"] * scales["R0"] * scales["a"]
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, eq_params, sheet_params=None, constants=None):
+        """Compute boundary force error.
+
+        Parameters
+        ----------
+        eq_params : dict
+            Dictionary of equilibrium degrees of freedom, eg Equilibrium.params_dict
+        sheet_params : dict
+            Dictionary of sheet current degrees of freedom.
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self.constants
+
+        Returns
+        -------
+        f : ndarray
+            Boundary force error (N).
+
+        """
+        if constants is None:
+            constants = self.constants
+        src_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._eq_data_keys,
+            params=eq_params,
+            transforms=constants["src_transforms"],
+            profiles=constants["src_profiles"],
+        )
+        eval_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._eq_data_keys,
+            params=eq_params,
+            transforms=constants["eval_transforms"],
+            profiles=constants["eval_profiles"],
+        )
+        if self._sheet_current:
+            assert sheet_params is not None
```

Comment:
is this a debug statement? Is it still needed? Shouldn't this get caught somewhere else first, like in the optimizer when it checks that the set of passed-in things is the same as obj.things, if the sheet current is not passed in with `things`?

Reply:
i think its still a useful check, in the event `obj.compute` is called by itself.